src/client/mod.rs

```diff
@@ -0,0 +1,592 @@
+            c.write_u32::<Endian>(0)?;
+            // record short and long help offset
+            let shorthelp_off_off = c.position();
+            c.write_u64::<Endian>(0)?;
+            let longhelp_off_off = c.position();
+            c.write_u64::<Endian>(0)?;
+
+            // value block
+            let value_block_off = self.wi.value_sec_off + i*VALUE_BLOCK_LEN;
+            c.set_position(value_block_off);
+            let mut string_val_off_off = None;
+            if type_code == STRING_METRIC_TYPE_CODE {
+                // value
+                c.write_u64::<Endian>(0)?;
+                // extra (string offset)
+                string_val_off_off = Some(c.position());
+                c.write_u64::<Endian>(0)?;
+            } else {
+                value_offset = c.position() as usize;
+                value_size = NUMERIC_VALUE_SIZE;
+
+                // value
+                m.write_val(&mut c)?;
+                // extra (string offset)
+                c.write_u64::<Endian>(0)?;
+            }
+            // offset to metric block
+            c.write_u64::<Endian>(metric_block_off)?;
+            // offset to instance block
+            c.write_u64::<Endian>(0)?;
+
+            // string block
+            let mut string_block_off = self.wi.string_sec_off
+                + self.wi.n_strings*STRING_BLOCK_LEN;
+
+            // short help
+            if m.shorthelp().len() > 0 {
+                c.set_position(shorthelp_off_off);
+                c.write_u64::<Endian>(string_block_off)?;
+
+                c.set_position(string_block_off);
+                c.write_all(CString::new(m.shorthelp())?.to_bytes_with_nul())?;
+
+                self.wi.n_strings += 1;
+                string_block_off += STRING_BLOCK_LEN;
+            }
+
+            // long help
+            if m.longhelp().len() > 0 {
+                c.set_position(longhelp_off_off);
+                c.write_u64::<Endian>(string_block_off)?;
+
+                c.set_position(string_block_off);
+                c.write_all(CString::new(m.longhelp())?.to_bytes_with_nul())?;
+
+                self.wi.n_strings += 1;
+                string_block_off += STRING_BLOCK_LEN;
+            }
+
+            // string value
+            match string_val_off_off {
+                Some(off_off) => {
+                    c.set_position(off_off);
+                    c.write_u64::<Endian>(string_block_off)?;
+
+                    c.set_position(string_block_off);
+                    m.write_val(&mut c)?;
+
+                    value_offset = string_block_off as usize;
+                    value_size = STRING_BLOCK_LEN as usize;
+
+                    self.wi.n_strings += 1;
+                },
+                None => {}
+            }
+
+            // update string count in string TOC block
+            c.set_position(self.wi.n_strings_off);
+            c.write_u32::<Endian>(self.wi.n_strings as u32)?;
+
+        }
+
+        { // set mmap_view for metric
```

Comment:
with blocks, it'd be cool if you also added short notes about lifetimes and why it was used at the beginning/end